clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

```diff
@@ -467,6 +469,583 @@ public void testClusterAuthorizationExceptionInInitProducerIdRequest() throws Ex
         assertSendFailure(ClusterAuthorizationException.class);
     }
 
+
+    @Test
+    public void testIdempotenceWithMultipleInflights() throws Exception {
```

Comment:
I was looking for a test case which exercised the sequence number adjustment logic, but couldn't find one. Specifically a test case which caused the sequence number of a batch to change.

Reply:
Good catch. There wasn't a test, and there was a bug in that logic. Particularly, when a batch fails fatally, the future in flight batches are guaranteed to fail with an `OutOfOrderSequenceException`. We retry these exceptions as long as the batch in question is not the next-in-line batch. However, when we reset sequence numbers, the first future batch does become the next-in-line batch, and it is not retried when it fails with --the expected-- `OutOfOrderSequenceException`. 

I have have the test case and the fix for the bug in an upcoming commit.